### PR TITLE
Better use of std::enable_if

### PR DESCRIPTION
**Pull request by beinhaerter**
Replace all occurances of `class = std::enable_if_t<Cond>` and `typename = std::enable_if_t<Cond>` with `std::enable_if_t<Cond, bool> = true`.

This commit is inspired by #1174, which changed one occurance in the owner header. This commit is aimed to fix all remaining occurances.

### CODE REVIEW

File: docs/headers.md
Comment (carsonRadtke) [inline docs/headers.md:103]: I am hesitant to accept these operator changes because I don't believe they are "dangerous". From my experiments it is impossible to override the `std::enable_if_t` with an arbitrary type.

If I'm wrong, could you give me an example that compiles where the shift is a non-integer type?

File: docs/headers.md
Comment (carsonRadtke) [inline docs/headers.md:137]: This is a change I am willing to accept because it makes the following example not compile:
```c++
#define GSL_USE_STD_BYTE 0
#include <gsl/gsl>
static_assert(std::is_floating_point_v<decltype(to_integer<float, float>(gsl::byte(0)))>, "oof");
```
(which [currently compiles](https://godbolt.org/z/1efd7sxsa) on `main`)

File: docs/headers.md
Comment (carsonRadtke) [inline docs/headers.md:242]: I can't figure out how to bypass any of these `enable_if_t`s. (Though I have little confidence my attempts were exhaustive).

Could you provide an example of why this change is necessary?

File: docs/headers.md
Comment (carsonRadtke) [inline docs/headers.md:313]: This is a change I am willing to accept. I missed this with my latest PR.

**Review by carsonRadtke — CHANGES_REQUESTED** [review]
I had originally considered asking for this change as part of #1174, but I was worried there may be subtle differences that need to be considered for each change.

From my point of view, I would like justification for each change (i.e. an example of accepted code that shouldn't be accepted). If the `std::enable_if_t` is impossible to override, I don't think the change is necessary. That being said, I will bring this PR up during the monthly maintainers sync (which happens to be this evening) to get a second opinion.

File: docs/headers.md
Comment (beinhaerter) [reply docs/headers.md:103]: I started my changes with byte and added tests in `TEST(byte_tests, bitwise_operations )` like this:
```
operator<< <float,void>(a, float{1});
operator>> <float,void>(a, float{1});
operator<<= <float,void>(a, float{1});
operator>>= <float,void>(a, float{1});
```
These look totally strange, and they do trigger a compile error, but the error is inside the operator in byte. With the change the compile error is at the call site (no matching operator found).

Comment (carsonRadtke) [reply docs/headers.md:103]: Ah! I didn't consider that. If these changes make compiler errors easier to understand, then I am all for it. I will re-evaluate with this point in mind.

File: docs/headers.md
Comment (beinhaerter) [reply docs/headers.md:242]: I'll try to come up with one example per change.

**Comment by beinhaerter** [comment]
I did some investigation. It seems not be possible to explicitely choose the template instantiation for a templated constructor in a class template, so my changes for `not_null` and `span` were not useful, I removed them.

For the other changes I split my PR into three commits:
1. add the tests that should fail but that do not fail
2. do the code changes that make the tests fail as expected
3. revert all changes from the first commit

I would have preferred to keep the tests from the first commits, but my SFINAE skills are too low. I tried it the way `span_compatibility_tests` is doing it, but I did not get it working.

**Comment by beinhaerter** [comment]
I am now happy with the PR. I was able to write tests like `static_assert(XXCompilesFor<YY>, "");` that work for the code before the fix and where negating the condition like `static_assert(!XXCompilesFor<YY>, "");` works after the fix. So we can now keep the tests in place and get rid of the third commit that deleted the tests.

**Comment by carsonRadtke** [comment]
Thanks for adding examples and it is awesome that they are now part of the tests. I will take another look once the tests are green for all compilers.

FYI - I made a change in the repo settings; the actions should now automatically run for your PRs.

**Comment by carsonRadtke** [comment]
This looks great! Thanks for making all those changes.

**Pull request merged**
